### PR TITLE
New api proposal

### PR DESCRIPTION
**Pull request by gyzerok**
[Draft document](https://gist.github.com/gyzerok/e8bbd6e074b195b62ce1b23a65affca7)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -36,6 +25,74 @@ and what they will do to your environment.
...
+      // Enter a sequence at FIRST_ACTION, SECOND_ACTION and THIRD_ACTION will be
+      // dispatched in the order they are passed to batch
+      return loop(
+        state.set('firstRun', true),
+        Effects.batch([
+          Effects.promise(doSecondAction, 'hello'),
+          Effects.constant(thirdAction)
+        ])
+      );
+      return {
```
Comment (fobos) [inline README.md]: Looks like you trying to return twice from one the case statement

**Comment by jevakallio** [comment]
Awesome work @gyzerok!

I have a few suggestions I would like to make. I really like the direction of reducing the amount of Effects methods. However, the main method `Effects.fromLazyPromise` is a very specific use case, and as your example shows, if people want to simply fire off an action without doing any fancy async stuff, they need to create both a thunk and a resolved promise, which feels quite inconvenient.

I'd like to propose a more flexible API with the following semantics:

If user passes an object, the object is treated as an action and dispatched
```js
Effects.create({ type: 'foo' });
```

If user passes a thunk that returns an object, the object is treated as an action and dispatched
```js
Effects.create(() => ({ type: 'foo' })
```

If thunk returns a Promise, the promise is resolved. If the promise returns an
action, the action is dispatched
```js
Effects.create(() => Promise.resolve({ type: 'foo' });
```

If user passes an array or a thunk that returns an array, they are all dispatched in serial. This should also work with promises, so that we can use `Promise.all` to execute multiple async
side effects and dispatch their results
```js
Effects.create([{ type: 'foo' }, { type: 'bar' }]);
Effects.create(() => [{ type: 'foo' }, { type: 'bar' }]);
Effects.create(() => Promise.all([
  Promise.resolve({ type: 'foo' }), 
  Promise.resolve({ type: 'bar' })
]);
```

If thunk returns nothing, that's totally fine. The side effect can be treated as a sink.
Sometimes you just want to cause a side effect and not dispatch another action.
From a developer experience there is a slight risk that a dev can forget to return
an action and a bug can go unnoticed, so this could also be a separate method,
e.g. Effects.sink
```js
Effects.create(() => {
  doSideEffect()
});
```

That's the basic semantics. I don't think `Effects.create` is a good name, but maybe something that is explicitly not very "FP style" to help with wider adoption. E.g.
 * `Effects.then`
 * `Effects.dispatch`
 * `Effects.next`

Thoughts? /cc @lukewestby

**Comment by jevakallio** [comment]
One thing I forgot to add was that if the create function is passed more than one argument, the ...rest args should be passed to the thunk as arguments. This enables one of the best features of redux-loop v2, namely that async effects can be tested by comparing a function reference and arcs for equality!

**Comment by lukewestby** [comment]
I would like to see the `Effects` namespace go away altogether and just have an `effect` function

```
import Api from 'api'
import { effect } from 'redux-loop'

effect(() =>
  Api.getStuff(arg)
    .then(successAction, failureAction))
```

If we're going for a small API and more approachability to JS then having `Effects` be a thing with a namespace like that seems to me to give the impression that it's a bigger deal than it is.

Initially I'd like to advocate against having the effect creator function being more flexible because it allows you to do things that are in general bad for being able to reason about your reducer. For example, I think being able to pass in a bare action to be called next leads to really confusing relationships with actions causing other actions in order to reuse their handlers when really what needs to be done is some function of the state needs to be extracted out of the reducer. I feel that something that happens inside of the effect creator function should actually be an effect in the literal sense. That said, we should talk about this some more maybe on a different issue and figure out if it makes sense to, for example, allow both async effects that return a promise for an action, and sync effects that return an action but might throw. I'm not a hard no on this one.

Having `.map` as an instance method on an effect works for me, and I also like having the returned effects from the reducer always be in an array to remove the batching concept from the API.

I'm totally cool with keeping the extra args for deep equality testing of effects. Personally I've picked up other techniques for testing logic around effects from working with Elm and Haskell but it was really nice to have for testing when working on a team that is focused on JS.

Lastly, I'm going to have to vote against effects that don't return an action for the exact reason @jevakallio mentioned: a dev might forget to return something and instead of seeing an error they'll accidentally be using a different feature. The behavior behind `Effects.sink` is essentially identical to an ordinary redux middleware as well so if people feel this is missing our docs can just point them  toward writing a middleware.

**Comment by lukewestby** [comment]
@gyzerok @jevakallio you should both be able to create branches in `redux-loop/redux-loop` now (let me know if you don't!), so in the future let's have PRs and discussions like this one over there to keep everything in one place

**Comment by gyzerok** [comment]
@lukewestby 
Actually I can transfer all off my changes to the main repo and create PR there. However I am not sure we want a lot of attention from people watching repository right now. It would be way better to share work when it is somewhat completed.

Now when I thought a bit more about your idea of having just single function for effect creation it makes a lot of sense to me. And I guess it would make much more sense to the average JS dev to have `map` as an instance method. Don't know why I haven't spotted it before!

@jevakallio
Thank you for your suggestions! After todays evening feel really awesome that we have even more people interested in new release ;)

Regarding discussion in general I'd say that I agree with what Luke already said. And I'll try to give you another perspective on this.

The idea behind this API proposal was to reuse reasoning behind Elm development. Shortly the idea is that whenever you can do something with existing set of APIs you don't need to blow up your API surface area to create more abstractions then necessary. Most of the time it can have not as positive effect as you might think which I'll try to show you later.

There are a few ways you can do everything you described.

**You can call `reducer` recursively**

Whenever you need to run one action immediately after another you can just call `reducer` again providing it with this specific action. This is not an obvious things and I see all that time that new Elm developers just don't think about it.

```js
function reducer(state, action) {
  case 'FIRST_ACTION':
    const { state: nextState } = reducer(state, { type: 'SECOND_ACTION' });
    return {
      state: nextState,
      effects: []
    };
}
```

Here I guess comes Luke's point that it's way easier to reason about nested function calls then to reason about whole Redux dispatch loop.

Perhaps we can just elaborate this idea in some kind of *tips & tricks* section in the docs.

**You can create API specific to your project**

If anyway you want to return action as affect from your reducer you pretty much can do it with the current API. Here I agree with you that it is inconvenient. And what you can do is create your own convenient function which will abstract this complexity away.

```js
function effect(action) {
  return Effects.fromLazyPromise(() => Promise.resolve(action));
}
```

So no you can use it and be happy :)

This might be good candidate for the docs as well.

**We can add new API to the library**

This also possible. And one thing here to consider is that it comes with several trade-offs.

1. It increases learning curve.
2. It encourages developers to use less nice ways of organising their code. Again a reference to Luke's response.
3. It adds significant maintainability cost to the library. We need to maintain more complex code, more tests, more documentation and examples.
4. And here can work some kind of theory of broken windows. If we already just simply added some convenience method why won't we and another one and one more?

What we will end up with if we will address each specific issue with new API?

**Conclusion**

Personally I would prefer to elaborate with this particular issue in that particular order I described things. Last option affects all the consumers and maintainers of the library and thus should be a very rare case.

In Elm you need to have really motivating use case for making things more complex. And most of the time the idea is that you simply can't do some sort of things with existing API. I'd love if we can use this exact reasoning while working together on this library!

What do you think about this? Have first two options addressed your necessities?

**Comment by jevakallio** [comment]
@lukewestby @gyzerok thanks for the considered feedback! I fully appreciate the reasoning of the limited API surface and think it's a great goal to have. 

I've considered it, and come to agree that I was being too maximalist with my initial draft and didn't think it through properly. While sinks are convenient, they can be achieved without explicit support, and I've never found multiple effects (aka. `Effects.batch`) too useful anyway.

I do, however, think that the `() => Promise<Action>` signature will not cover one key use case that I have found invaluable while writing apps with `redux-loop`, which is using vanilla action creators to be used as Effect creators.

If we simply supported `(...args) => Action | Promise<Action>`, we would enable this pattern. I hope I can convince you of this and we can land in a compromise 😉 

## Deep equality testing of effects is a key feature

In my experience, and in discussions with other members of the community, one of the big reasons developers go out and seek more "advanced" async/effects middleware is testability. Enabling developers to test the returned effects by simply comparing them to expected effects is an amazing feature and should not be crippled. A lot of the reasoning below will be motivated by this.

(That said, @lukewestby I would love to hear about "other techniques for testing logic around effects from working with Elm and Haskell" you mentioned, but that's a topic for another day.)

## Dispatching synchronous actions solves coupling between siblings in nested store

We often use `combineReducers` as a mechanism to modularize our state management. Consider the following state structure:
```js
combineReducers({
  navigation: navigationReducer,
  login: loginReducer
  //...many more
});
```

We want to separate the concerns between navigation state (which in this case is passed to React Native navigator) and our login flow that consists of multiple screens. However, in a login flow we need to be able to cause "effects" across the subtrees of the state:
 
1. Update the login flow state 
2. Push a new route on the navigation route stack

A simplified example on how we'd use redux-loop here:
```js
//navigation.js
export const pushRoute = (route, routeParams) => ({ 
  type: 'Navigation/PUSH_ROUTE', 
  payload: { route, routeParams }
});
```

```js
//login.js
function loginReducer(state, action) {
  switch(action.type) {
    case 'USERNAME_ENTERED':
      return {
        state: { ...state, username: action.payload },
        effect: effect(pushRoute, 'login/password', { email: state.loginType })
      }
  }
}
```

Solving this without `redux-loop` poses a challenge for us, because:
 * Because `combineReducers` has passed only the `login` subtree into our reducer, we cannot update the other `navigation` subtree from `loginReducer` (nor would we want to, for spaghetti)
 * Because we don't have access to `navigation` state subtree, we cannot recursively call `navigationReducer` with the new action (nor would we want to, for excessive direct coupling)
 * Because `navigationReducer` (not included in code snipped) does not have access to `login` subtree, it cannot handle the 'USERNAME_ENTERED' action directly, because it needs `store.login.loginType`.

Now, you could argue that this is a problem only because we are using `combineReducers`, and instead we should solve the issue by refactoring our state or extracting a shared, higher-order reducer that can update the entire state. And we *could*, but we would lose the modularity we use combineReducers for in the first place. Dispatching an effect via `redux-loop` is such a nice way of handling this.

Semantically, I think this maps very well to the concept of Effects. Instead of causing an effect in an external system over I/O, we are causing a direct effect in another subsystem. To me this is a really clear mental model of the relationships between my data, and helps formalize that relationship. And most importantly, to test the effects between these subsystems instead of relying on magical behaviour.

## Back to testability

So, in the current proposal, the developer developer cannot use the `pushRoute` action creator directly. She either 
 * needs to `Promise.resolve` inline in the reducer and sacrifice the equality testing of the effect, or 
 * create a separate, public `pushRouteEffect` method that lifts the action into a promise, which is semantically unclear.

It is a very small API surface increase from `() => Promise<Action>` to `(...args) => Action | Promise<Action>`, but it makes the library much more useful and intuitive to use.

## Fin.

I eagerly await your thoughts :)

**Comment by gyzerok** [comment]
### Some thoughts

> Now, you could argue that this is a problem only because we are using combineReducers, and instead we should solve the issue by refactoring our state or extracting a shared, higher-order reducer that can update the entire state.

Sorry, that's exactly what I gonna do :D

> Dispatching synchronous actions solves coupling between siblings in nested store

Does it? It looks like they are still retain coupling, but the it becomes more vague and implicit. Putting coupled things under single reducer is exactly what Redux propose you to do. One of the reasons for that is that you do not have access to one's reducer state in another.

> And we could, but we would lose the modularity we use combineReducers for in the first place. Dispatching an effect via redux-loop is such a nice way of handling this.

Just think about it a bit. If you split one of your modules into multiple files which have each one line of code would it make you code more modular? Sometimes the problem you are trying to solve is complex enough. Most of the time you just can't make your solution less complex then your problem is. You can of course hide this complexity behind implicitness and abstractions, but it won't remove it, just make it less discoverable.

Thus the problem you describe is not anyhow related to Redux or Redux Loop, but is the case of your particular way of designing your state. It doesn't mean that your way is good or bad and what's more important it doesn't mean that one of these libraries should be designed to address this problem.

> Semantically, I think this maps very well to the concept of Effects. Instead of causing an effect in an external system over I/O, we are causing a direct effect in another subsystem. To me this is a really clear mental model of the relationships between my data, and helps formalize that relationship. And most importantly, to test the effects between these subsystems instead of relying on magical behaviour.

This sounds to me a bit like microservice architecture. And this is totally okay to design your system like this. You should just be prepared to address trade-offs which it brings with itself.

In Elm Architecture there is no way to `composeReducers` generically like you do it. The whole idea is about composing them manually by nesting one piece of state within another. What that mean is that reducer nesting mirrors view nesting. In this case you can explicitly pass some parts of global state as a context for child reducers. For example if you have auth token you need to put it in all of your requests. The most explicit way, I believe, is to pass it down as context to your child reducers which then could run HTTP requests using it.

### There is still way to solve it

One other thing regarding your example is the fact that navigation is indeed side-effectful. Somewhere there should be call to `navigator.push`. If we consider this, than your code will be something like the following:

```js
//login.js
function loginReducer(state, action) {
  switch(action.type) {
    case 'USERNAME_ENTERED':
      return {
        state: { ...state, username: action.payload },
        effect: effect(() => new Promise(resolve => navigator.push({ route: 'login/password' }))
      }
  }
}
```

Since this code is side-effectful you can not guarantee that it will be executed synchronously. Even if it will do it right now it may be changed in React Native under the hood at any time so why would we rely on it?

So in this case even if you want to design your state in a way you do it right now there is a solution without adding new API method which I described earlier.

```js
function createConstantEffect(action) {
  return Effects.fromLazyPromise(() => Promise.resolve(action));
}

function loginReducer(state, action) {
  switch(action.type) {
    case 'USERNAME_ENTERED':
      return {
        state: { ...state, username: action.payload },
        effect: createConstantEffect(pushRoute)
      }
  }
}
```

### Conclusion

It seems that you might feel as we are a bit stubborn on this case, I am sorry for that. And I pretty much can understand you because this particular reasoning gave me a lot of bad feelings in Elm community in the beginning. The thing is - after a while I started to understand it and enjoy as you hopefully will if we can convince you :)

So in our case options would look like this:
1. People who want to follow a way of designing application state as you described address trade-offs of side-stepping Elm architecture by themselves.
2. Since library is so small anyone can actually make own fork with changes one find useful.
3. We change library API to support specific use case which Elm Architecture originally do not support.

One of the things here is that 2 option is totally fine. And when I had my first call with @lukewestby about API proposal I've said that I totally ready if he discard all of my thought and I will just make my custom fork for my custom usage. It appeared that they do not violate any ideas behind the library, so we decided that better way would be to contribute to the original one :)

**Comment by jevakallio** [comment]
@gyzerok well, that's all fine and dandy. There are many ways to structure the application state, each with their tradeoffs. I have personally found the structure I describe to work well. You may have other experiences.

As a library, `redux-loop` can be as opinionated as it wants. Each opinion it prescribes limits how it can be used. But fundamentally it's a library that will be used in different environments and contexts, not only in a context of fractal Elm architecture.

If you zoom out a bit, all I'm really proposing here is to allow the effect creator to return an action, as well as a promise of an action. 

It's trivial to implement, and doesn't add a single method to the library. In fact, it's not even proposing to *add anything*. This behaviour already exists as `Effects.call`. I'm proposing just not to remove an existing feature that has proven to be useful in practice when writing real life applications.

You propose that if people want to add higher level functionality, they can fork the library. They can, but I don't believe that's a good idea. Each app out there with their own homegrown fork is wasteful from maintenance perspective, and doesn't do anything to help the credibility and adoption of this library.

There is really no logical argument to be made here for either option. Either you can support synchronous actions officially, or you can force people to hack synchronous actions with pointless Promise.resolve.

@lukewestby's call I guess.

**Comment by gyzerok** [comment]
@jevakallio 

> As a library, redux-loop can be as opinionated as it wants. Each opinion it prescribes limits how it can be used. But fundamentally it's a library that will be used in different environments and contexts, not only in a context of fractal Elm architecture.

Actually the library is descendant of Elm Architecture in it's essence. Of course with my proposal we side-step it. And if you look at it we do not side-step the general idea, only implementation detail which make more sense for being used in JS environment by JS devs.

The thing here is that you on one hand want Elm Architecture way of doing things and on another hand don't want it. It sounds like a clear point to implement some custom library. If someone comes to Redux and says that he want a mechanism of building dependencies between reducers because he organize his code in some different way, should the team implement it (as there was in other Flux options)?

I would say that one of the core principles of Elm Architecture is that it to certain level dictates you how to do things. I would personally rephrase it as "it teaches you how to do things".

> You propose that if people want to add higher level functionality, they can fork the library.

Maybe order of my comments might be a bit misleading. Not necessarily.

I propose that if people want some higher order functionality they can create helpers for that on top of the library. This will move complexity of maintaining this higher order functionality to their application. And since not all of the people need this higher order functionality I won't say that it's responsibility of the library to maintain it. And people who don't need this higher order functionality don't get it by default, which is ok. And this I cover with the solution of having helper function to create action effects in your own codebase. Do you see that there is something wrong with it?

If this higher order functionality is impossible to implement by your own then we have to options: try to address it within the library or admit that this do not follow library idea and you can implement your own version.

One example for this are Redux middlewares. You can put each and every one in Redux, but not all the people need them. That's why they do exist as a separate libraries. Even though DevTools are used by majority of developers they are not part of the core library.

> Either you can support synchronous actions officially, or you can force people to hack synchronous actions with pointless Promise.resolve.

What is more poinless:
1. Support functionality within the library which are not important in order to use library
2. Build your specific way of writing code on top of general purpose library

Again I'd like to put your attention to the fact that you **can** achieve similar functionality by having one small function within your application for that purpose. It's not something impossible to do with current API idea.

> There is really no logical argument to be made here for either option.

Why doesn't being able to achieve the same thing without increasing API doesn't sound logical to you?

------------------------------

Have you had a chance to use Elm? It is usual thing there that people initially start fighting with Elm Architecture. Just give it a try :)

**Comment by jevakallio** [comment]
@gyzerok 

> Again I'd like to put your attention to the fact that you can achieve similar functionality by having one small function within your application for that purpose.

I assume you are referring to this code snippet from above:
```js
function createConstantEffect(action) {
  return Effects.fromLazyPromise(() => Promise.resolve(action));
}
```

It it only were so simple, I'd happily drop this. However, having a generic utility function breaks the testability of the effect. The anonymous function created inside this helper will never be equal to anything that we can create in our test code, so we can't assert for its equality.

To do this, we need to create a public, named function for each action creator we want to lift to use as an effect:

```
const fooCreator = () => ({ type: 'foo' });
const fooEffect = () => Promise.resolve(fooCreator());

const barCreator = () => ({ type: 'bar' });
const barEffect = () => Promise.resolve(barCreator());
```

So that we can assert that the returned effect `===` the effect creator function.

To be charitable, API-wise, I can implement some kind of helper:
```
const fooCreator = () => ({ type: 'foo' });
const fooEffect = lift(fooCreator);
```

But I'll still need to do this for each action type. Which is fine. But let's not pretend that it's just a single line of code. It's `n` lines of code for `n` effects I want to support, plus the additional indirection of not being able to just fire an action without having to modify code in both at the Effect call site and where the action is created.

> Why doesn't being able to achieve the same thing without increasing API doesn't sound logical to you?

It would, but it's not the case here. We are not achieving the same thing.

**Comment by gyzerok** [comment]
@jevakallio 

> However, having a generic utility function breaks the testability of the effect.

Oh sorry, I missed this point. How would you do comparison of HTTP request side-effects look like?

On the other hand if we are talking about just returning plain actions (not actual side-effects), then you can test it as follows:

```js
const { state, effect } = reducer(/* ... */);

effect.toPromise().then(action => assert(action, expectedAction));
```

So you can still use this function while preserving a way to test nicely.

**Comment by jevakallio** [comment]
@gyzerok testing HTTP request side-effects and sync side effects should look the same:
```js
  expect(effect.promiseCreator).to.be(fooActions.someAsyncHttpActionCreator);
  expect(effect.args).to.be(state.userId);

  expect(effect.promiseCreator).to.be(barActions.someSyncActionCreator);
  expect(effect.args).to.be(state.userId);
```

In this case, the `promiseCreator` would be renamed to something else, of course, since it would not necessarily be a _promise_ creator.

**Comment by gyzerok** [comment]
@jevakallio it seems the way of testing async effects you outlined in the last post won't work with current nor with API you propose. What I am missing here?

In order to continue discussion on the topic we might want to have clear and rather verbose answers to the following questions:
1. What is the reason we want to be able to write tests for side-effects? How actual use case where it help should look like?
2. What is expected ROI of these tests?
3. What kind of feedback do we want to get from the tests for side-effects?
4. How this problem is being solved in other libraries/languages?
5. While having answers to previous questions how can we imagine best way to write this tests form the end-user perspective?

Feel free to correct/add questions.

**Comment by gyzerok** [comment]
Notes, TODO:
- [x] Better error messages (polyfills, exceptions …)
- [x] Types (TypeScript, Flowtype)
- [ ] Examples (Elm architecture)
- [ ] Add missing API documentation
- [ ] Migration path
- [ ] A bit change effects by providing equality check and eliminating namespace

**Pull request closed without merging**